src/tracer.c

```diff
@@ -117,7 +117,66 @@ find(pid_t pid)
     return pinfo + i;
 }
 
-uint8_t        *hash_file(char *);
+uint8_t *hash_file(char *);
+
+static void
+handle_close(int fd, FILE_INFO *f)
+{
+    f->hash = hash_file(f->path);
+}
+
+static void
+handle_open(pid_t pid, int syscall, int fd, const unsigned long long *args)
+{
+
+    if (fd >= 1024)		       // more than 1024 files open
+	// concurrently
+    {
+	error(EXIT_FAILURE, errno,
+	      "limit of 1024 open files exceeded for process %d", pid);
+    }
+
+    PROCESS_INFO *pinfo = find(pid);
+    FILE_INFO *finfo = next_finfo(pinfo);
```

Comment:
That won't work.
Since we now expect the info to be in the specific `fd` index of the array, `next_finfo()` should never be used.

Reply:
@zvr We don't?

Isn't `finfo` supposed to be an array that holds all files ever opened? Making it so that `finfo[fd] = file` isn't  a valid option since `fd`s are recycled as previously mentioned.

This is also the purpose of `open_files`, an array that maps an `fd` to the relative `FILE_INFO` index at `finfo`.

in other words, instead of having 
`finfo[fd] = file`
 we have
 `finfo[open_files[fd]] = file`
 
 Doesn't this solve our problem? Am i missing something?

Reply:
Why making our lives difficult?

Simply use `finfo[fd]` and do not even have an `open_files`.

Otherwise a number of other decisions are not valid; for example why have only 1024 open_files? Compiling the Linux kernel will definitely use many more fies.

There is no problem with recycling of open fd's, since we will processing/recording every file on `close()` -- not after program execution.

Reply:
@zvr 1024 are the max files that can be open concurrently.

But problem is we are storing the hashes in the `finfo` array, if we each time overwrite the `FILE_INFO` at index `fd` we will be losing data. If the limit of 1024 isn't enough i can see three solutions:

1. Increase `open_files` array size appropriately
2. Make `open_files` a dynamic array
3. Have a global `FILE_INFO` array that holds the already processed files. In that sense after a file is `closed` we will be moving it from one array to the other.

The third solution matches better to what you are describing, "recording every file on `close()`". And i agree that it will be easier than using `open_files`. So give me a thumbs up if i can go on and implement this solution.

Reply:
@zvr review this version, i think it's what you wanted.